src/sdo/data_loader/goes/cache.py

```diff
@@ -0,0 +1,152 @@
+
+def last_day_of_month(date):
+    if date.month == 12:
+        return date.replace(day=31).day
+    last_ts = date.replace(month=date.month+1, day=1) - dt.timedelta(days=1)
+    return last_ts.day
+
+
+def get_date_ranges(start, end):
+    # retrieve monthly ranges between start and end to not overload the HEK API
+    # https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases
+
+    adjusted_start = dt.datetime(start.year, start.month, 1, 0, 0, 0)
+    adjusted_end = dt.datetime(
+        end.year, end.month, last_day_of_month(end), 23, 59, 59)
+
+    start_dates = pd.date_range(start=adjusted_start, end=adjusted_end,
+                                freq="MS").to_pydatetime().tolist()
+
+    end_dates = pd.date_range(start=adjusted_start, end=adjusted_end,
+                              freq="M").to_pydatetime().tolist()
+
+    if len(start_dates) == 0:
+        return [(adjusted_start, adjusted_end)]
+
+    ranges = []
+    for s, e in zip(start_dates, end_dates):
+        ranges.append((s, e))
+
+    if len(ranges) == 0:
+        return [(adjusted_start, adjusted_end)]
+
+    return ranges
+
+
+def fetch_goes_metadata(start, end, path):
+    # NOTE it can happen that sunpy will download invalid files during rate limiting, find them using grep -r "Too many" ~/sunpy/data/*
+    # https://docs.sunpy.org/en/v3.0.0/generated/gallery/acquiring_data/goes_xrs_example.html
+    # https://www.ngdc.noaa.gov/stp/satellite/goes-r.html
+    # https://www.ngdc.noaa.gov/stp/satellite/goes/index.html
+
+    # Since March 2020, data prior to GOES 15 (incl) is no longer supported by NOAA and GOES 16 and 17 data is now provided.
+    # GOES 16 and 17 are part of the GOES-R series and provide XRS data at a better time resolution (1s).
+    # GOES 16 has been taking observations from 2017, and GOES 17 since 2018
+    for t_start, t_end in get_date_ranges(start, end):
+        goes_dir = Path(path) / Path("goes_cache") / Path(str(t_start.year)) /\
+            Path(str(t_start.month))
+        goes_path = goes_dir / Path('goes_ts.csv')
+        if not Path.exists(goes_path):
+            try:
+                logger.info(f"retrieving goes data for {goes_path}")
+                os.makedirs(goes_dir, exist_ok=True)
+                search_result = Fido.search(
+                    a.Time(t_start, t_end), a.Instrument.xrs)
+
+                logger.info(f"found {len(search_result)} search results")
+                for result in search_result:
+                    # https://docs.astropy.org/en/stable/table/operations.html#grouped-operations
+                    grouped = result.group_by("Start Time")
+                    rows_to_remove = np.array([], dtype=int)
+                    for group in grouped.groups:
+                        if len(group) > 1:
+                            # only retain the observations for the newest known satellite
+                            satellite_number = np.max(group["SatelliteNumber"])
```

Comment:
for each day this will prefer the newest satellite that is available and not download data for older Satellites

Reply:
*Disclaimer* bit of a HACK but I did not find a better way to work with Astropy tables